blackjax/adaptation/mclmc_adaptation.py

```diff
@@ -0,0 +1,243 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Algorithms to adapt the MCLMC kernel parameters, namely step size and L.
+
+"""
+
+from typing import NamedTuple
+
+import jax
+import jax.numpy as jnp
+from jax.flatten_util import ravel_pytree
+
+from blackjax.diagnostics import effective_sample_size  # type: ignore
+from blackjax.util import pytree_size
+
+
+class MCLMCAdaptationState(NamedTuple):
+    """Represents the tunable parameters for MCLMC adaptation.
+
+    Attributes:
+        L (float): The momentum decoherent rate for the MCLMC algorithm.
+        step_size (float): The step size used for the MCLMC algorithm.
+    """
+
+    L: float
+    step_size: float
+
+
+def mclmc_find_L_and_step_size(
+    kernel,
+    num_steps,
+    state,
+    part1_key,
+    part2_key,
+    frac_tune1=0.1,
+    frac_tune2=0.1,
+    frac_tune3=0.1,
+):
+    """
+    Finds the optimal value of L (step size) for the MCLMC algorithm.
+
+    Args:
+        kernel: The kernel function used for the MCMC algorithm.
+        num_steps: The number of MCMC steps that will subsequently be run, after tuning
+        state: The initial state of the MCMC algorithm.
+        frac_tune1: The fraction of tuning for the first step of the adaptation.
+        frac_tune2: The fraction of tuning for the second step of the adaptation.
+        frac_tune3: The fraction of tuning for the third step of the adaptation.
+
+    Returns:
+        state: The final state of the MCMC algorithm.
+        params: The final hyperparameters of the MCMC algorithm.
+    """
+    dim = pytree_size(state.position)
+    params = MCLMCAdaptationState(jnp.sqrt(dim), jnp.sqrt(dim) * 0.25)
+    varEwanted = 5e-4
+
+    state, params = make_L_step_size_adaptation(
+        kernel=kernel,
+        dim=dim,
+        frac_tune1=frac_tune1,
+        frac_tune2=frac_tune2,
+        varEwanted=varEwanted,
+        sigma_xi=1.5,
+        num_effective_samples=150,
+    )(state, params, num_steps, part1_key)
+
+    if frac_tune3 != 0:
+        state, params = make_adaptation_L(kernel, frac=frac_tune3, Lfactor=0.4)(
+            state, params, num_steps, part2_key
+        )
+
+    return state, params
+
+
+def make_L_step_size_adaptation(
+    kernel,
+    dim,
+    frac_tune1,
+    frac_tune2,
+    varEwanted=1e-3,
+    sigma_xi=1.5,
+    num_effective_samples=150,
+):
+    """Adapts the stepsize and L of the MCLMC kernel. Designed for the unadjusted MCLMC"""
+
+    gamma_forget = (num_effective_samples - 1.0) / (num_effective_samples + 1.0)
+
+    def predictor(state_old, params, adaptive_state, rng_key):
```

Comment:
state_old -> previous_state
state_new -> next_state

Same for below and other files.

Reply:
Reminder to change the naming.